### PR TITLE
New labs

### PR DESCRIPTION
**Pull request by palandreassen**

### CODE REVIEW

File: getting-started.rst [hunk 1]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
-=====
+- Create a pipe with ``"_id": "order-anonymized-labs"``.
+- Set source to be of type "dataset" and the dataset to "webshop-order".
+- In the transform section, copy all properties from source except the ``customerId`` and ``customerId-ni`` attributes.
+- Start the pipe and check the results in the "Output"-tab of the pipe. 
+
+Tips
+
+  - You can create a downstream pipe (a pipe that uses the dataset of the pipe you're currently viewing as its source) from the pipe menu. Click the three dots to the right of the pipe name to open the menu.
+  - The templates avaiable on the pipe's "Config"-tab can help ease and speed up config creation. Add a transform config with the "Add DTL transform" button.
```
Comment (TarjeiSkrede) [inline getting-started.rst:1627]: Minor typo (avaiable).

File: getting-started.rst [hunk 2]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+    * ``["copy", "foo*"]`` will copy all attributes that have "foo" at the start of their name.
 
 .. _getting-started-labs-2:
 
-Scope: Remove and add.
+Lab 2
+^^^^^
+
+| **Add and remove attributes**
+| **Variables _S and _T**
```
Comment (TarjeiSkrede) [inline getting-started.rst:1643]: Are descriptions missing here?

File: getting-started.rst [hunk 3]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+++++
+
+- Create an input pipe with ``_id`` "geonorge-county" that imports county data using the newly made "geonorge"-system. 
+- Give the pipe a JSON source config and set its ``url`` to "kommuneinfo/v1/fylkerkommuner". 
+- As there is no automatically identifiable ID in the data you will have to add a transform config in your pipe and in it ``add`` the ``_id`` attribute from the source data's "fylkesnummer" attribute (county number).
+- ``copy`` all attributes from the source.
+
+Tips
+
+  - You can find the JSON source documented under Service Configuration in the Developer Guide section of the ducumentation.
```
Comment (TarjeiSkrede) [inline getting-started.rst:1721]: Minor type (ducumentation).

File: getting-started.rst [hunk 4]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
-=====
 .. _getting-started-labs-8:
 
-Scope : Publishing data to CSV-endpoint
+Lab 8
+^^^^^
 
-  * Create new pipe. Source from lab 6. Add transform and sink.​
+| **Filtering entities**  
+| **Global dataset as pipe source**
```
Comment (TarjeiSkrede) [inline getting-started.rst:1778]: Descriptions missing here too?

File: getting-started.rst [hunk 5]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+1. Create a pipe with **global-person** as the source dataset. Do not include a transform section in the pipe. 
+2. Observe that the pipe outputs all the same entities as **global-person** in its "Output"-tab.
+3. Add a transform to the pipe config with ``["copy, "*"]`` and ``["filter", ["in", "~:hr:person", "_S.rdf:type"]]``. 
+4. Restart the pipe. Observe that there are now only 10 entities in the output, and that they all contain data from "hr-person".
+5. Change the filter to keep only entities that have data from the **crm-person** dataset.
+6. Restart the pipe and observe that you now have 10 entities which all have data from **crm-person**. This should not be the exact same 10 entities as before.
+
+Tips
+
+  - The entities you find in the "Output"-tab of the pipe are the same as the entities in the dataset with the same name. The entities in the "Latest" category is the current dataset. In "Latest w/ deleted" you will also find the hr-person entities som the first pipe run. In the "All" category you will find previous versions of all entities as well as the current and deleted entities.
```
Comment (TarjeiSkrede) [inline getting-started.rst:1804]: Som should be from?

File: getting-started.rst [hunk 6]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+Lab 12
+^^^^^^
+
+| **Casting**
+| **Number logic**    
+| **Nested functions**
+
+We can cast an attribute as different types, like integer, string, datetime and others described in the docs. 
+
+When working with numbers we have a set of functions and operators that we can apply to the numeric values. These are documented in the DTL Reference Guide under "Math". Different functions for casting can be found under "Numbers", "Strings" and other headers regarding data types.
```
Comment (TarjeiSkrede) [inline getting-started.rst:1895]: Should we be consistent in when we link to the DTL documentation?

File: getting-started.rst [hunk 7]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+- Imagine a new order has come in for everyone in this dataset, and increase the number of ``TimesOrdered`` by one. As this attribute is a string in the source data it needs to be cast to integer before we can increase the value. Cast the result back to string and add it to the entity as a new ``TimesOrdered`` attribute which will get its namespace from the new pipe.
+
+.. _getting-started-labs-13:
+
+Lab 13
+^^^^^^
+
+**String logic**
+
+In Sesam's DTL there are many functions available for working with :ref:`strings <string_dtl_function>`. Let's try working wuth some of them.
```
Comment (TarjeiSkrede) [inline getting-started.rst:1922]: Typo: wuth

File: getting-started.rst [hunk 8]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+  - The first two digits of a SSN represent the day, the next two digits represent the month and then the next two represent the year.
+
+.. _getting-started-labs-14:
+
+Lab 14
+^^^^^^
+
+**Datetime logic**
+
+Sesam provides a toolkit for working with date and time values, just like it does for string and number logic. These DTL-functions can be found in the docs too! With ``datetime-parse"`` we can translate strings to datetime values, while ``datetime-format`` translates datetime values to strings. ``datetime-diff``, ``datetime-plus`` and ``datetime-shift`` lets us do calculation and manipulation of datetime values.
```
Comment (TarjeiSkrede) [inline getting-started.rst:1946]: There is an " in ``datetime-parse"``

File: getting-started.rst [hunk 9]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+Lab 8
+^^^^^
 
-  * Create new pipe. Source from lab 6. Add transform and sink.​
+| **Filtering entities**  
+| **Global dataset as pipe source**
+| **Pipe reset**
 
-Hint: Look :ref:`here <getting-started-csv-endpoint>` for help
+As explained in our :ref:`Best Practices <best-practice>` we want to use a global dataset as our data source when preparing data for export to another system. When we use a global dataset as a pipe's source we have to specify what data from that global dataset we want to use. We use the DTL transform function ``filter``, to select only the entities we want. The most common filtering we do is filter on ``rdf:type``, an attribute we add to the data as it enters Sesam to tag which system it comes from and what it contains. A filter on rdf:type can look like this:
```
Comment (TarjeiSkrede) [inline getting-started.rst:1781]: Maybe change "We use the DTL transform function.." to "For this, we use the DTL transform function..".

File: getting-started.rst [hunk 10]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+.. _getting-started-labs-16:
+
+Lab 16
+^^^^^^
+
+**Filter**
+
+There are two ``filter`` functions in Sesam's DTL. In this task and when we filter entities on ``rdf:type`` we use the transform function. Transform functions are the functions that can add or remove attributes to the entities, which in this case adds an internal system attribute (``"_filtered": true``) to the entity. The other filter function is found under Values/Collections in the docs and is used to filter values that do not match a certain criteria when we perform logic on data within a transfom function.
+
+If we use the ``filter`` transform function without any input parameter we filter out the entitiy we're currently processing unconditionally. We can provide a function as an "unless"-condition, and the entity will only be filtered if that function returns true. We have already used this filter function to filter entities from a global dataset source based on ``rdf:type``. ``["filter", ["in", "~\:hr:person", "_S.rdf:type"]]`` reads "Filter, unless '~:hr:person' is in the 'rdf:type'-list".
```
Comment (TarjeiSkrede) [inline getting-started.rst:1993]: Typo: entitiy

File: getting-started.rst [hunk 11]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+- In hte DTL transform first add ``["filter", ["neq", "_S._deleted", true]]`` to remove ``_deleted`` entities, as the emit children transform would also emit children from those. 
+- As the municipalities listed in the counties do not have an "_id" already we need to add that before put them in ``$children``. In the DTL transform have a ``create-child`` transform function take the results of an ``apply`` as its argument. The ``apply`` function should pass the list-attribute "_S.kommuner" (municipalities) through a rule in which we build our new entities. In that rule 
+
+  - copy all attributes that are passed to the rule in the "kommuner" list object with ["copy", "*"]
+  - add "_id" from "_S.kommunenummer" (municipality number)
+  - add "rdf:type" as ["ni", "geonorge", "municipality"] to use for filtering downstream when we are to use this data later.
+- Run the pipe and confirm that it outputs municipalities.
+- Finaly, add the new municipality dataset to **global-location**'s' source dataset list and run the global pipe.
+
+Tips
```
Comment (TarjeiSkrede) [inline getting-started.rst:2078]: Maybe add a tip about dependency tracking when it comes to 'create', 'create-child' and 'emit-children'.

File: getting-started.rst [hunk 12]
```diff
@@ -1598,151 +1590,2033 @@ When communicating with the API we use requests methods such as **GET**, **POST*
...
+
+The counties we imported from geonorge earlier in these labs each include a list of municipalities. Let's emit these municipalities as their own entities using ``create-child`` and an emit children transform that we chain after the DTL transform where we create the ``$children``.
+
+Task
+++++
+
+- Make a pipe downstream from geonorge-county and name it geonorge-municipality.
+- Give it a chained transform with a DTL transform first and an emit children transform second. 
+- In hte DTL transform first add ``["filter", ["neq", "_S._deleted", true]]`` to remove ``_deleted`` entities, as the emit children transform would also emit children from those. 
+- As the municipalities listed in the counties do not have an "_id" already we need to add that before put them in ``$children``. In the DTL transform have a ``create-child`` transform function take the results of an ``apply`` as its argument. The ``apply`` function should pass the list-attribute "_S.kommuner" (municipalities) through a rule in which we build our new entities. In that rule
```
Comment (TarjeiSkrede) [inline getting-started.rst:2070]: As the municipalities listed in the counties do not have an "_id" already, we need to add that before we put them in ``$children``.

Comment (grove) [reply getting-started.rst:2070]: Typo: In hte DTL

File: getting-started.rst (hunk 2 above)
Comment (palandreassen) [reply getting-started.rst]: I have added some more descriptive text to this lab

File: getting-started.rst (hunk 4 above)
Comment (palandreassen) [reply getting-started.rst]: Have added some more descriptive text to this lab as well

**Pull request closed without merging**